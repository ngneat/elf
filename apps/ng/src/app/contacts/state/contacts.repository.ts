import { Injectable } from '@angular/core';
import { createStore } from '@ngneat/elf';
import { addEntities, withEntities } from '@ngneat/elf-entities';
import {
  PaginationData,
  selectCurrentPage,
  selectCurrentPageEntities,
  selectPaginationData,
  setCurrentPage,
  setPage,
  updatePaginationData,
  withPagination,
} from '@ngneat/elf-pagination';

interface Contact {
  id: number;
  email: string;
  name: string;
  address: string;
}

const store = createStore(
  { name: 'contacts' },
  withPagination(),
  withEntities<Contact>()
);

@Injectable({ providedIn: 'root' })
export class ContactsRepository {
  activePage$ = store.pipe(selectCurrentPage());
  paginationData$ = store.pipe(selectPaginationData());
  activePageContacts$ = store.pipe(selectCurrentPageEntities());

  setActivePage(id: Contact['id']) {
    store.update(setCurrentPage(id));
  }

  addContacts(page: number, contacts: PaginationData & { data: Contact[] }) {
    const { data, ...pagination } = contacts;

    store.update(
      addEntities(data),
      updatePaginationData(pagination),
      setPage(
        page,
        data.map((c) => c.id)
      )
    );
  }

  get store() {
    return store;
  }
}
