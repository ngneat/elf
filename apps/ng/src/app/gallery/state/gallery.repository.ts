import { Injectable } from '@angular/core';
import { createStore } from '@ngneat/elf';
import {
  addEntities,
  selectAllEntities,
  selectEntities,
  UIEntitiesRef,
  unionEntities,
  updateEntities,
  withEntities,
  withUIEntities,
} from '@ngneat/elf-entities';

export interface GalleryItem {
  id: number;
  title: string;
  description: string;
  image: string;
}

export interface GalleryUIItem {
  id: number;
  open: boolean;
}

const store = createStore(
  { name: 'gallery' },
  withEntities<GalleryItem>(),
  withUIEntities<GalleryUIItem>()
);

@Injectable({ providedIn: 'root' })
export class GalleryRepository {
  items$ = this.store
    .combine({
      entities: store.pipe(selectAllEntities()),
      UIEntities: store.pipe(selectEntities({ ref: UIEntitiesRef })),
    })
    .pipe(unionEntities());

  addGalleryItems(galleryItems: GalleryItem[]) {
    const uiGalleryItems = galleryItems.map(({ id }) => ({ id, open: false }));
    store.update(
      addEntities(galleryItems),
      addEntities(uiGalleryItems, { ref: UIEntitiesRef })
    );
  }

  get store() {
    return store;
  }

  toggleItemOpen(galleryItemId: number) {
    this.store.update(
      updateEntities(galleryItemId, (item) => ({ ...item, open: !item.open }), {
        ref: UIEntitiesRef,
      })
    );
  }
}
