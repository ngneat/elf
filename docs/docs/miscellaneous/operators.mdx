# Operators

Elf provides the following operators:

## filterNil
Filters `undefined` or `null` values:

```ts
import { filterNil } from '@ngneat/elf';

foo$.pipe(filterNil());
```

## head
Get the the first item from an array:

```ts
import { head } from '@ngneat/elf';

skills$.pipe(head());
```

## distinctUntilArrayItemChanged

A `distinctUntillChanged` implementation for arrays:

```ts
import { distinctUntilArrayItemChanged } from '@ngneat/elf';

skills$.pipe(distinctUntilArrayItemChanged());
```