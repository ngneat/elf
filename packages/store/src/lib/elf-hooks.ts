import { NotVoid } from './types';

interface ElfHooksRegistry {
  preStoreUpdate?: (
    currentState: any,
    nextState: any,
    storeName: string
  ) => NotVoid<any>;
}

// this is internal object that's not exported to public API
export const elfHooksRegistry: ElfHooksRegistry = {};

class ElfHooks {
  registerPreStoreUpdate<T>(
    fn: (currentState: any, nextState: any, storeName: string) => NotVoid<T>
  ) {
    elfHooksRegistry.preStoreUpdate = fn;
  }
}

export const elfHooks = new ElfHooks();
