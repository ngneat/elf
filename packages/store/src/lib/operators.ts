import {
  asapScheduler,
  MonoTypeOperatorFunction,
  OperatorFunction,
  pipe,
} from 'rxjs';
import {
  debounceTime,
  distinctUntilChanged,
  filter,
  map,
} from 'rxjs/operators';

export function select<T, R>(mapFn: (state: T) => R): OperatorFunction<T, R> {
  return pipe(map(mapFn), distinctUntilChanged());
}

export function head<
  T extends any[],
  Item = T extends (infer I)[] ? I : never
>(): OperatorFunction<T, Item> {
  return map((arr: T) => arr[0] as Item);
}

export function distinctUntilArrayItemChanged<T>(): MonoTypeOperatorFunction<
  T[]
> {
  return distinctUntilChanged((prevCollection: T[], currentCollection: T[]) => {
    if (prevCollection === currentCollection) {
      return true;
    }

    if (prevCollection.length !== currentCollection.length) {
      return false;
    }

    const isOneOfItemReferenceChanged = currentCollection.some((item, i) => {
      return prevCollection[i] !== item;
    });

    // return false means there is a change and we want to call next()
    return !isOneOfItemReferenceChanged;
  });
}

export const asap = <T>(): MonoTypeOperatorFunction<T> =>
  debounceTime(0, asapScheduler);

export function filterNil<T>(): OperatorFunction<T, NonNullable<T>> {
  return filter(
    (value: T): value is NonNullable<T> => value !== null && value !== undefined
  );
}
