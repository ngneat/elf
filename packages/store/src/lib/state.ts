type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never;
type Merge<State extends any[], Key extends PropertyKey> = UnionToIntersection<
  State[number][Key]
>;

export type PropsFactory<Props, Config> = { props: Props; config: Config };
export type EmptyConfig = undefined;

export function createState<S extends PropsFactory<any, any>[]>(
  ...propsFactories: S
): { state: Merge<S, 'props'>; config: Merge<S, 'config'> } {
  const result = { config: {}, state: {} };

  for (const { config, props } of propsFactories) {
    Object.assign(result.config, config);
    Object.assign(result.state, props);
  }

  return result as { state: Merge<S, 'props'>; config: Merge<S, 'config'> };
}
