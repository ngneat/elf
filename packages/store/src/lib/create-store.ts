import { createState, PropsFactory } from './state';
import { Store } from './store';

type UnionToIntersection<U> = (U extends any ? (k: U) => void : never) extends (
  k: infer I
) => void
  ? I
  : never;
type Merge<State extends any[], Key extends PropertyKey> = UnionToIntersection<
  State[number][Key]
>;

export function createStore<
  S extends [PropsFactory<any, any>, ...PropsFactory<any, any>[]]
>(
  storeConfig: StoreConfig,
  ...propsFactories: S
): Store<{
  name: string;
  state: Merge<S, 'props'>;
  config: Merge<S, 'config'>;
}> {
  const { state, config } = createState(...propsFactories);
  const { name } = storeConfig;

  return new Store({ name, state, config });
}

export interface StoreConfig {
  name: string;
}
