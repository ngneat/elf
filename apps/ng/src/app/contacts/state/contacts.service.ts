import { Injectable } from '@angular/core';
import { getData } from '../contacts-data';
import { timer } from 'rxjs';
import { map, tap } from 'rxjs/operators';
import { ContactsRepository } from './contacts.repository';
import { skipWhilePageExists } from '@ngneat/elf-pagination';

@Injectable({ providedIn: 'root' })
export class ContactsService {
  constructor(private repo: ContactsRepository) {}

  getContacts(page: number) {
    return timer(300).pipe(
      map(() => getData({ page })),
      tap((res) => this.repo.addContacts(page, res)),
      skipWhilePageExists(this.repo.store, page)
    );
  }
}
