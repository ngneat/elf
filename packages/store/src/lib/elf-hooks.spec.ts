import { createStore } from './create-store';
import { elfHooks, elfHooksRegistry } from './elf-hooks';
import { select } from './operators';
import { withProps } from './props.state';

describe('elfHooks', () => {
  describe('METHOD: registerPreStoreUpdate', () => {
    it('preStoreUpdate should be called if defined', () => {
      elfHooks.registerPreStoreUpdate(() => {
        return { filter: 'elf' };
      });

      const store = createStore(
        { name: 'todos' },
        withProps<{ filter: string }>({ filter: '' })
      );

      const spy = jest.spyOn(elfHooksRegistry, 'preStoreUpdate');

      store.update((s) => ({ ...s, filter: 'foo' }));

      expect(spy).toHaveBeenCalledTimes(1);
      expect(spy).toHaveBeenCalledWith(
        { filter: '' },
        { filter: 'foo' },
        'todos'
      );

      expect(store.getValue()).toMatchSnapshot();
    });

    it('registerPreStoreUpdate should be properly typed', () => {
      // @ts-expect-error - should prohibit void functions
      elfHooks.registerPreStoreUpdate(() => {
        console.log('void function');
      });
    });

    it('subscriber should emit only once', () => {
      elfHooks.registerPreStoreUpdate((_, currentState) => {
        return { num: currentState.num + 1 };
      });

      const store = createStore(
        { name: 'counter' },
        withProps<{ num: number }>({ num: 0 })
      );

      const spy = jest.fn();

      store.pipe(select((s) => s.num)).subscribe(spy);

      store.update((s) => ({ ...s, num: s.num + 1 }));

      expect(spy).toHaveBeenCalledTimes(2);
      expect(store.getValue().num).toBe(2);
    });
  });
});
