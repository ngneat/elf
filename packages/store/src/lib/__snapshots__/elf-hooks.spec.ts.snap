// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`elfHooks METHOD: registerPreStoreUpdate preStoreUpdate should be called if defined 1`] = `
Object {
  "filter": "elf",
}
`;
